### PR TITLE
Fix fmt V11 build errors 

### PR DESCRIPTION
**Pull request by yingsu00**
Recently brew updated fmt version from 10.1 to 11.0. In order to prevent build errors from happening after we upgrade Folly, the following changes are made:

1. Error "'this' argument to member function 'format' has type 'const fmt::formatter<facebook::velox::exec::Spiller::Type>', but function is not marked const".
The new fmt v11 forces the const check. To fix this error, the "const" keyword was added to the `format()` function declarations.

2. Error "error: static assertion failed due to requirement 'formattable_char': Mixing character types is disallowed." The new fmt v11 added a check if the arguments of fmt::make_format_args() have mixing different character types (e.g., char vs wchar_t or std::string vs folly::Range<const char*>). To fix this, we need to convert the folly::StringPiece to string.

These changes will be needed after we upgrade Folly, which is also dependent on fmt v11. The changes are backward compatible for fmt v10.

Fixes https://github.com/facebookincubator/velox/issues/10886

### CODE REVIEW

File: CMakeLists.txt
Comment (zuyu) [inline CMakeLists.txt:414]: nit: I don't think we need to specify the version any more.

File: scripts/setup-macos.sh
Comment (zuyu) [inline scripts/setup-macos.sh:139]: Please also remove the `function install_fmt` implementation in the same file.

File: velox/connectors/hive/HiveDataSource.cpp
Comment (zuyu) [inline velox/connectors/hive/HiveDataSource.cpp:19]: nit for header file ordering: move it under STL header files.

File: velox/exec/tests/TableScanTest.cpp
Comment (zuyu) [inline velox/exec/tests/TableScanTest.cpp:21]: nit for header file ordering:

```
#include <atomic>
#include <shared_mutex>

#include <fmt/ranges.h>
#include <folly/synchronization/Baton.h>
#include <folly/synchronization/EventCount.h>
#include <folly/synchronization/Latch.h>
```

Note that also fix `#include "folly/experimental/EventCount.h"` to `#include <folly/synchronization/EventCount.h>`

Comment (yingsu00) [reply velox/exec/tests/TableScanTest.cpp:21]: Thanks for the reminder. There are other headers to be cleaned up. Will send a separate PR for that.

File: velox/connectors/hive/HiveDataSource.cpp
Comment (yingsu00) [reply velox/connectors/hive/HiveDataSource.cpp:19]: @zuyu In Velox, the corresponding .h shall be put at the beginning in its own group. See the dicussion in https://github.com/facebookincubator/velox/issues/1918

File: velox/exec/tests/TableScanTest.cpp
Comment (yingsu00) [reply velox/exec/tests/TableScanTest.cpp:21]: Just FYI, I personally think the include order should be local to global:

Imagine you are writing a library that depends on some other library. You might do something like this:

```
#include <vector> //STL
#include <gtest/gtest.h> //ThirdParty 
#include "geometry.hpp" //In-house
 
#define PI 3.2 //As pr. Indiana Bill #246, 1897
 
TEST(TestGeometry, rotatingOrigoGivesOrigo) {
    std::vector<double> v(2,0);
    rotate(v, PI);
    EXPECT_EQ(0, v[0]) << "X got moved!";
    EXPECT_EQ(0, v[1]) << "Y got moved!";
}
```
and your geometry.hpp looks like this:

```
void rotate(std::vector<double>& v, double angle);
```
This all works out nicely until someone else wants to include geometry.hpp without including <vector> first, and get something like geometry.hpp:1: error: ‘vector’ is not a member of ‘std’. **You have now forced all the users of your geometry library to include <vector> before including geometry.hpp**.

But the maintainers think this is less an issue and no resource to update all include headers now.

Comment (zuyu) [reply velox/exec/tests/TableScanTest.cpp:21]: I have the similar preference as https://github.com/facebookincubator/velox/issues/1918#issuecomment-1172541369, as well as the Google code style for the header ordering https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes.

I did not see much difference as your proposed style.

**Comment by yingsu00** [comment]
@Yuhta @assignUser @kgpai Will you please review this PR? I cannot request reviews recently.

**Review by majetideepak — CHANGES_REQUESTED** [review]
fmt version should be fixed to avoid such breakages from sudden package manager updates.
We should install fmt and not use the brew version.

@czentgr Any thoughts?

**Comment by czentgr** [comment]
@majetideepak I agree. Using brew installs as dependency for the Velox/Prestissimo build is a problem. We would be forced in these emergency upgrades which are risky. We should decouple. homebrew should only install utilities used for the build and not dependencies we need for the build to succeed. However, the way that the build currently works we always pick up linked installations from homebrew. Linux is much more conservative when it updates and usually doesn't change major versions as often.

In this particular case, we need to ensure Linux also uses the same version of fmt. We also don't know if prestissimo will work on with these updates or if it will be broken.
In the short term we should keep the versions installed in the system using the `install` functions and ensure that the build is using these as we control them.


However, longer term we should look at upgrading dependencies eventually but without the pressure of failing builds. So these changes ares till useful.

File: CMakeLists.txt
Comment (assignUser) [reply CMakeLists.txt:414]: It's a minimum version so we should keep it. The issue is that fmt doesn't provide a proper version config that would prevent us from picking up an incompatible version.

**Review by assignUser — COMMENTED** [review]
Thank for introducing the changes. I do agree with @majetideepak and @czentgr that we should 'pin' the fmt version we use as they seem to release more major versions with breaking changes than not... I still think we should add the cpp changes (afaik they should be backwards compatible?) as we will want to bump folly soon'ish and then we will need it anyway.

**Comment by zuyu** [comment]
@majetideepak In general, I agree that we should fix the versions of dependency libraries, and upgrade the libraries when necessary. `brew`, however, is never good at installing a fixed version of a given library, as a given version might be removed from `brew`. If we choose to live with `brew`, we have to upgrade libraries often, or only using it for certain libraries.

IMO, some dependency libraries (e.g., `folly`) should be latest, or at least latest enough. And if earlier we upgrade `folly` to at least `v2024.08.05.00` from `v2024.05.20.00`, `fmt` upgrade via `brew` would not break the build.

In other words, this PR should be treated as a `folly` upgrade PR, and we should also do it for other boxes (e.g., Linux), not only for macOS.

**Comment by yingsu00** [comment]
> Thank for introducing the changes. I do agree with @majetideepak and @czentgr that we should 'pin' the fmt version we use as they seem to release more major versions with breaking changes than not... I still think we should add the cpp changes (afaik they should be backwards compatible?) as we will want to bump folly soon'ish and then we will need it anyway.

This PR is actually bumping up folly version for MacOS. I will add Linux upgrade too, but I don't have a box right now to test it.
And yes, the C++ changes are backward compatible.

**Comment by yingsu00** [comment]
> IMO, some dependency libraries (e.g., folly) should be latest, or at least latest enough. And if earlier we upgrade folly to at least v2024.08.05.00[1](https://github.com/facebookincubator/velox/pull/10904#user-content-fn-1-5d72cfe226068dba8da9cfbd5952c557) from v2024.05.20.00, fmt upgrade via brew would not break the build.

Yes that's right. Since folly is using a fixed version and a specific install_folly function, its dependencies should also be fixed snapshot version. Shall we change the way to install them from brew to its own install_xxx functions as the next step?

In this PR, I will restore install_fmt and fix it at 11.0.2, but will not touch other dependencies.

**Comment by assignUser** [comment]
> In this PR, I will restore install_fmt and fix it at 11.0.2, but will not touch other dependencies.

In light of #10670 that's a good idea :)

**Comment by majetideepak** [comment]
As @czentgr suggested, I am unsure/not confident about upgrading folly without proper testing on all the platforms.
We also decided to keep other Meta packages such as fbthrift, wangle, etc. that depend on FB_OS_VERSION to be same.
I am working on a fix to pick installed versions over brew versions here https://github.com/facebookincubator/velox/pull/10920

**Comment by majetideepak** [comment]
@kgpai what is the internal FB_OS_VERSION Meta uses? Are you always on the latest release?

**Comment by yingsu00** [comment]
Yes it seems there are more work needed to upgrade Folly: https://github.com/facebookincubator/velox/pull/10924
I have updated this PR only for the C++ changes, and will close https://github.com/facebookincubator/velox/pull/10924 and leave the Folly upgrade to capable hands. Please review again. @majetideepak @assignUser @czentgr @zuyu

**Comment by kgpai** [comment]
@majetideepak Yes internally FB_OS_VERSION (i.e folly) is always latest .

**Comment by zuyu** [comment]
> @majetideepak Yes internally FB_OS_VERSION (i.e folly) is always latest .

@kgpai It's good to know that. So do you suggest `velox` also to use the latest `FB_OS_VERSION` when necessary, i.e., this fmt 11 build issue?

**Comment by yingsu00** [comment]
fmt build failures happening: https://github.com/facebookincubator/velox/actions/runs/10712752127/job/29703663400?pr=10898 
```
/opt/homebrew/include/fmt/base.h:1627:3: error: static assertion failed due to requirement 'formattable_char': Mixing character types is disallowed.
  static_assert(formattable_char, "Mixing character types is disallowed.");
  ^             ~~~~~~~~~~~~~~~~
/opt/homebrew/include/fmt/base.h:2018:20: note: in instantiation of function template specialization 'fmt::detail::make_arg<true, fmt::context, folly::Range<const char *>, 0>' requested here
  return {{detail::make_arg<NUM_ARGS <= detail::max_packed_args, Context>(
                   ^
/opt/homebrew/include/fmt/format.h:4365:28: note: in instantiation of function template specialization 'fmt::make_format_args<fmt::context, folly::Range<const char *>, 1UL, 0UL, 15ULL, 0>' requested here
  return vformat(fmt, fmt::make_format_args(args...));
                           ^
/Users/runner/work/velox/velox/_build/Release/_deps/folly-src/folly/IPAddressV6.cpp:218:14: note: in instantiation of function template specialization 'fmt::format<folly::Range<const char *> &>' requested here
        fmt::format("Invalid input. Got '{}'", piece));
             ^
1 error generated.
```

**Comment by alirezaraha3** [comment]
> fmt build failures happening: https://github.com/facebookincubator/velox/actions/runs/10712752127/job/29703663400?pr=10898 
> 
> ```
> 
> /opt/homebrew/include/fmt/base.h:1627:3: error: static assertion failed due to requirement 'formattable_char': Mixing character types is disallowed.
> 
>   static_assert(formattable_char, "Mixing character types is disallowed.");
> 
>   ^             ~~~~~~~~~~~~~~~~
> 
> /opt/homebrew/include/fmt/base.h:2018:20: note: in instantiation of function template specialization 'fmt::detail::make_arg<true, fmt::context, folly::Range<const char *>, 0>' requested here
> 
>   return {{detail::make_arg<NUM_ARGS <= detail::max_packed_args, Context>(
> 
>                    ^
> 
> /opt/homebrew/include/fmt/format.h:4365:28: note: in instantiation of function template specialization 'fmt::make_format_args<fmt::context, folly::Range<const char *>, 1UL, 0UL, 15ULL, 0>' requested here
> 
>   return vformat(fmt, fmt::make_format_args(args...));
> 
>                            ^
> 
> /Users/runner/work/velox/velox/_build/Release/_deps/folly-src/folly/IPAddressV6.cpp:218:14: note: in instantiation of function template specialization 'fmt::format<folly::Range<const char *> &>' requested here
> 
>         fmt::format("Invalid input. Got '{}'", piece));
> 
>              ^
> 
> 1 error generated.
> 
> ```
> 
> 

[_****_]()

**Comment by majetideepak** [comment]
It is unlikely that Meta is using fmt 11 internally due to this build issue. @kgpai can you share the versions of the various meta packages (folly, fizz, wangle, mvfst, fbthrift) used internally? 
I have a fix for the fmt build failure on MacOS here https://github.com/facebookincubator/velox/pull/10920

I am not confident we should land this PR since it will hide the situation where fmt11 headers are being picked instead of the fmt10 version. A failure is expected in this situation.

**Comment by yingsu00** [comment]
@majetideepak This change is backward compatible and will unblock the CI failures. It is needed anyways when Folly is upgraded. Why can't we merge it? When fmt is updated in the future, no error is supposed to be expected. This is to prevent the error from happening, not hiding it. Why would adding some const keywords make anything wrong? Even though fmt is still v10, the const keywords make the code safer. And adding the toString() to folly::Stringpiece has no harm. Why do we want to complicate future Folly upgrade and redo all the work when the errors happen in the future?

**Comment by kgpai** [comment]
FB_OS_VERSION and fmt version is different; Internally we still only use fmt 9 ; I dont think we are ready for 11 yet, unless its fully backwards compatible with 9.

**Comment by majetideepak** [comment]
@yingsu00, we have issues with our build. We see the brew includes being picked up over the installed packages.
This means that the fmtv11 headers could be used for fmtv10 library. This is not okay even if the build is succeeding.
I have a quick potential fix for the CI failure here https://github.com/facebookincubator/velox/pull/10933
The brew include problem is being resolved here https://github.com/facebookincubator/velox/pull/10920

**Comment by yingsu00** [comment]
> @yingsu00, we have issues with our build. We see the brew includes being picked up over the installed packages. This means that the fmtv11 headers could be used for fmtv10 library. This is not okay even if the build is succeeding. I have a quick potential fix for the CI failure here #10933 The brew include problem is being resolved here #10920

I know what you are saying and I saw your PRs. I agree those PRs, which fix the fmt version to be v10, should be merged to solve the current build errors. But Velox is going to upgrade Folly soon and that will also require fmt v11 upgrade. You will see the errors then and you will have to redo the same change here.  In case you didn't see it, I have updated this PR to only include the C++ changes that will be needed by then. These changes are backward compatible and can only make the code safer. So I think all https://github.com/facebookincubator/velox/pull/10670 #10920  #10933 and this PR should be merged, such that we stay on fmt v10 for now, and have a smooth upgrade on folly in the near future. Can you explain more about why this PR should not be merged? What is the risk?

@assignUser @kgpai What are your thoughts?

**Comment by assignUser** [comment]
I would agree that we should merge the source fixes for fmt 11, as long as they are backwards compatible with 10 (which they seem to be).

**Comment by yingsu00** [comment]
> I would agree that we should merge the source fixes for fmt 11, as long as they are backwards compatible with 10 (which they seem to be). (but maybe after the ci stuff has been merged and is ✔️ again

Thanks @assignUser! The tests are all green now, but sure we can wait until the other CI change PRs to be merged first.

**Comment by majetideepak** [comment]
There is no risk other than I find it odd to merge fmtv11 changes without officially supporting it.
Why not bump the fmt version to 11 in this PR if we need it for folly?

**Comment by yingsu00** [comment]
> There is no risk (after we fix the build issues) other than I find it odd to merge fmtv11 changes without officially supporting it. Why not bump the fmt version to 11 in this PR if we need it for folly?

fmt 11.0.2 does not work with the current version of folly, because of similar reasons:
```
In file included from /Users/yingsu/repo/presto1/presto/presto-native-execution/folly/folly/IPAddressV6.cpp:17:
In file included from /Users/yingsu/repo/presto1/presto/presto-native-execution/folly/folly/IPAddressV6.h:37:
In file included from /Users/yingsu/repo/presto1/presto/presto-native-execution/folly/folly/FBString.h:34:
In file included from /usr/local/include/fmt/format.h:41:
/usr/local/include/fmt/base.h:1627:3: error: static assertion failed due to requirement 'formattable_char': Mixing character types is disallowed.
  static_assert(formattable_char, "Mixing character types is disallowed.");
  ^             ~~~~~~~~~~~~~~~~
/usr/local/include/fmt/base.h:2018:20: note: in instantiation of function template specialization 'fmt::detail::make_arg<true, fmt::context, folly::Range<const char *>, 0>' requested here
  return {{detail::make_arg<NUM_ARGS <= detail::max_packed_args, Context>(
                   ^
/usr/local/include/fmt/format.h:4365:28: note: in instantiation of function template specialization 'fmt::make_format_args<fmt::context, folly::Range<const char *>, 1UL, 0UL, 15ULL, 0>' requested here
  return vformat(fmt, fmt::make_format_args(args...));
                           ^
/Users/yingsu/repo/presto1/presto/presto-native-execution/folly/folly/IPAddressV6.cpp:218:14: note: in instantiation of function template specialization 'fmt::format<folly::Range<const char *> &>' requested here
        fmt::format("Invalid input. Got '{}'", piece));
             ^
```

**Review by majetideepak — APPROVED** [review]
@yingsu00 thanks for verifying.
We need to update folly first then. But these changes can land.

**Review by assignUser — APPROVED** [review]

**Review by czentgr — APPROVED** [review]

**Comment by kgpai** [comment]
@yingsu00 Can you rebase and fix the merge conflict ? Thank you !

**Pull request closed without merging**
